### PR TITLE
fix: Fix resolveNamedType cannot resolve type parameter in contextual element

### PR DESCRIPTION
**Pull request by HerrCai0907**
fix: #2447 and #2446

- [x] I've read the contributing guidelines
- [x] I've added my name and email to the NOTICE file

### CODE REVIEW

File: src/resolver.ts [hunk 1]
```diff
@@ -188,8 +188,13 @@ export class Resolver extends DiagnosticEmitter {
...
     if (isSimpleType) {
       let simpleName = nameNode.identifier.text;
+      let type: Type | null = null;
       if (ctxTypes && ctxTypes.has(simpleName)) {
-        let type = assert(ctxTypes.get(simpleName));
+        type = assert(ctxTypes.get(simpleName));
+      } else {
+        type = ctxElement.lookupGenericType(simpleName);
+      }
+      if (type) {
```
Comment (dcodeIO) [inline src/resolver.ts:197]: This looks like there is a foregoing issue, as `ctxTypes` should have all the types. Is there perhaps something missing instead where `ctxTypes` is populated? Like, right now, `lookupGenericType` does after-the-fact what should already have been done before it seems.

File: src/resolver.ts [hunk 2]
```diff
@@ -763,7 +768,7 @@ export class Resolver extends DiagnosticEmitter {
           }
           let defaultType = typeParameterNode.defaultType;
           if (defaultType) {
-            let resolvedDefaultType = this.resolveType(defaultType, ctxFlow.actualFunction, contextualTypeArguments, reportMode);
+            let resolvedDefaultType = this.resolveType(defaultType, prototype, contextualTypeArguments, reportMode);
```
Comment (dcodeIO) [inline src/resolver.ts:771]: This is somewhat similar. The `actualFunction` should be an instance of `prototype`, then sharing state, or is it different?

Comment (HerrCai0907) [reply src/resolver.ts:771]: actualFunction is forcaller side.

File: src/resolver.ts (hunk 1 above)
Comment (HerrCai0907) [reply src/resolver.ts:197]: Yes, we just need to find a position to sync generic type in `ctxElement` to `ctxType` when `ctxElement` is `Function` or `Class`.

**Comment by HerrCai0907** [comment]
```typescript
class F<T> {
  foo(a: StaticArray<string>): string[] {
    const result = a.slice(0);
    // push some things onto "result"
    return result;
  }
}
export function foo(a: StaticArray<string>): string[] {
  let f = new F<f64>();
  return f.foo(a)
}
```

consider this case. In old version, generic `T` in class `F` will pass through `contextualTypeArguments` and cause error.

**Comment by dcodeIO** [comment]
I think I see now. Can you add a comment along the lines of...?

```
// Default parameters are resolved in context of the called function, not the calling function
```

**Review by dcodeIO — APPROVED** [review]

**Comment by MaxGraey** [comment]
Thanks!

**Pull request merged**